README.md
A library for parsing timecoded lyrics data from SYLT-frames encoded in mp3 files. (SYnchronized Lyrics/Text)

## Features

- Parse mp3 as byte list into an intermediate `SyltLyricsData` object.
- Export the `SyltLyricsData` as SRT or LRC lyrics text.

## Getting started

- Add the package to your `pubspec.yaml`.
- No further configuration is required.

## Usage

```dart
// Create a sample SyltLyricsData object and convert it to bytes. This list of
// bytes would normally be loaded from an mp3 file.
List<int> mp3Bytes = SyltLyricsData(
    textEncoding: TextEncoding.iso8859_1,
    language: 'eng',
    timestampFormat: TimestampFormat.absTime32BitMilliseconds,
    contentType: ContentType.lyrics,
    contentDescriptor: 'MiniLyrics',
    lyricsLines: [
    ('Hey', 255),
    ('Hey', 500),
    ('Hey', 512),
    ],
).toBytes();
// import 'dart:io';
// List<int> mp3Bytes = List<int> mp3Bytes = File('path/to/my/file.mp3').readAsBytesSync();

// Parse the list of bytes coming from the mp3 file.
// The library searches for the SYLT-frame within the ID3 tag and parses it,
// if one such frame is found.
SyltLyricsData lyricsData =
    SyltLyricsFromMp3Parser.parseMp3BytesToSyltLyricsData(mp3Bytes);

// The intermediate data can be parsed to either LRC or SRT lyrics formats:
String lrcLyrics = lyricsData.toLrc();
print('LRC lyrics:\n$lrcLyrics');

String srtLyrics = lyricsData.toSrt();
print('SRT lyrics:\n$srtLyrics');
```

## Additional information

- This package has mainly been developed to be used with my fork of the `christian_lyrics` package (https://github.com/Movinghead333/christian_lyrics). That package requires SRT lyrics for its functionality. These lyrics can be saved in a SYLT frame within the mp3. The `sylt_parser` package can then load the lyrics from the `.mp3` files directly. Both of those packages are used in my app `Flow Lyrix` (https://github.com/Movinghead333/flow_lyrix) which was developed as a replacement for MiniLyrics which does not seem to be working on newer Android versions.
- If you have any problems with the package, then feel free to file an issue.
- If you have feature request and suggestions, then file a pull-request.


CHANGELOG.md
## 0.2.0

- Added an exception in case no SYLT-frame can be found in a list of bytes to be parsed.
- Added documentation for public facing API methods and classes.

## 0.1.1

- When outputting to SRT format, blank lines are no replaced with a single space character so parsers working with empty lines as separators are less likely to fail, when parsing the output SRT lyrics.

## 0.1.0

- First version of the package with the core functionalities working.
- TODO: Add error handling


test/utility_test.dart
import 'package:sylt_parser/src/enums.dart';
import 'package:sylt_parser/src/utility.dart';
import 'package:test/test.dart';

void main() {
  group('Utility:', () {
    test('stringFromBytes() test 1', () {
      // Arrange
      // Create nullTerminated string with basic characters resulting in a utf8
      // conforming string
      List<int> bytes = [...'Hello World'.codeUnits, 0];

      // Act
      String parsedString =
          Utility.stringFromBytes(bytes, TextEncoding.iso8859_1);

      //Assert
      expect(parsedString, 'Hello World');
    });
  });
}


test/sylt_parser_test.dart
import 'package:sylt_parser/src/enums.dart';
import 'package:sylt_parser/src/sylt_lyrics_data.dart';
import 'package:sylt_parser/src/utility.dart';
import 'package:test/test.dart';

void main() {
  group('SyltLyricsData:', () {
    setUp(() {
      // Additional setup goes here.
    });

    test('toBytes() returns correct bytes list 1', () {
      // Arrange
      SyltLyricsData syltLyricsData = SyltLyricsData(
        textEncoding: TextEncoding.iso8859_1,
        language: 'eng',
        timestampFormat: TimestampFormat.absTime32BitMilliseconds,
        contentType: ContentType.lyrics,
        contentDescriptor: 'MiniLyrics',
        lyricsLines: [
          ('Hey', 255),
          ('Hey', 500),
          ('Hey', 512),
        ],
      );

      // Act
      List<int> bytes = syltLyricsData.toBytes();

      // Assert
      expect(bytes.length, 10 + 6 + 11 + 24);
      List<int> frameHeaderBytes = [83, 89, 76, 84, 0, 0, 0, 41, 0, 0];
      List<int> frameConfigBytes = [
        TextEncoding.iso8859_1.rawByte,
        ...Utility.languageStringToBytes('eng'),
        TimestampFormat.absTime32BitMilliseconds.rawByte,
        ContentType.lyrics.rawByte,
        ...Utility.stringToBytes('MiniLyrics', TextEncoding.iso8859_1)
      ];
      List<int> lyricsBytes = [
        ...Utility.stringToBytes('Hey', TextEncoding.iso8859_1),
        ...Utility.intToFourBytes(255),
        ...Utility.stringToBytes('Hey', TextEncoding.iso8859_1),
        ...Utility.intToFourBytes(500),
        ...Utility.stringToBytes('Hey', TextEncoding.iso8859_1),
        ...Utility.intToFourBytes(512),
      ];

      expect(bytes.sublist(0, 10), frameHeaderBytes);
      expect(bytes.sublist(10, 27), frameConfigBytes);
      expect(bytes.sublist(27), lyricsBytes);
    });

    test('toBytes() returns correct bytes list 2', () {
      // Arrange
      SyltLyricsData syltLyricsData = SyltLyricsData(
        textEncoding: TextEncoding.unicode,
        language: 'eng',
        timestampFormat: TimestampFormat.absTime32BitMilliseconds,
        contentType: ContentType.lyrics,
        contentDescriptor: 'MiniLyrics',
        lyricsLines: [
          ('Hey', 255),
          ('Hey', 500),
          ('Hey', 512),
        ],
      );

      // Act
      List<int> bytes = syltLyricsData.toBytes();

      // Assert
      int numLyricsBytes = 3 * (2 * 4 + 4);
      expect(bytes.length, 10 + 6 + 2 * 11 + numLyricsBytes);
    });
  });

  // group('A group of tests', () {
  //   final awesome = Awesome();

  //   setUp(() {
  //     // Additional setup goes here.
  //   });

  //   test('First Test', () {
  //     expect(awesome.isAwesome, isTrue);
  //   });
  // });
}


test/sylt_lyrics_data_test.dart
import 'package:sylt_parser/src/enums.dart';
import 'package:sylt_parser/src/sylt_lyrics_data.dart';
import 'package:sylt_parser/src/utility.dart';
import 'package:test/test.dart';

void main() {
  group('SyltLyricsData:', () {
    SyltLyricsData sampleSyltLyricsData = SyltLyricsData(
      textEncoding: TextEncoding.iso8859_1,
      language: 'eng',
      timestampFormat: TimestampFormat.absTime32BitMilliseconds,
      contentType: ContentType.lyrics,
      contentDescriptor: 'MiniLyrics',
      lyricsLines: [
        ('Hey', 255),
        ('Hey', 500),
        ('Hey', 512),
      ],
    );

    test('toBytes() returns correct bytes list 1', () {
      // Arrange
      SyltLyricsData syltLyricsData = sampleSyltLyricsData.copyWith();

      // Act
      List<int> bytes = syltLyricsData.toBytes();

      // Assert
      expect(bytes.length, 10 + 6 + 11 + 24);
      List<int> frameHeaderBytes = [83, 89, 76, 84, 0, 0, 0, 41, 0, 0];
      List<int> frameConfigBytes = [
        TextEncoding.iso8859_1.rawByte,
        ...Utility.languageStringToBytes('eng'),
        TimestampFormat.absTime32BitMilliseconds.rawByte,
        ContentType.lyrics.rawByte,
        ...Utility.stringToBytes('MiniLyrics', TextEncoding.iso8859_1)
      ];
      List<int> lyricsBytes = [
        ...Utility.stringToBytes('Hey', TextEncoding.iso8859_1),
        ...Utility.intToFourBytes(255),
        ...Utility.stringToBytes('Hey', TextEncoding.iso8859_1),
        ...Utility.intToFourBytes(500),
        ...Utility.stringToBytes('Hey', TextEncoding.iso8859_1),
        ...Utility.intToFourBytes(512),
      ];

      expect(bytes.sublist(0, 10), frameHeaderBytes);
      expect(bytes.sublist(10, 27), frameConfigBytes);
      expect(bytes.sublist(27), lyricsBytes);
    });

    test('toBytes() returns correct bytes list 2', () {
      // Arrange
      SyltLyricsData syltLyricsData = SyltLyricsData(
        textEncoding: TextEncoding.unicode,
        language: 'eng',
        timestampFormat: TimestampFormat.absTime32BitMilliseconds,
        contentType: ContentType.lyrics,
        contentDescriptor: 'MiniLyrics',
        lyricsLines: [
          ('Hey', 255),
          ('Hey', 500),
          ('Hey', 512),
        ],
      );

      // Act
      List<int> bytes = syltLyricsData.toBytes();

      // Assert
      int numLyricsBytes = 3 * (2 * 4 + 4);
      int numBytes = 10 + 6 + 2 * 11 + numLyricsBytes;
      expect(bytes.length, numBytes);
      List<int> frameHeaderBytes = [
        83,
        89,
        76,
        84,
        0,
        0,
        0,
        numBytes - 10,
        0,
        0
      ];
      List<int> frameConfigBytes = [
        TextEncoding.unicode.rawByte,
        ...Utility.languageStringToBytes('eng'),
        TimestampFormat.absTime32BitMilliseconds.rawByte,
        ContentType.lyrics.rawByte,
        ...Utility.stringToBytes('MiniLyrics', TextEncoding.unicode)
      ];
      List<int> lyricsBytes = [
        ...Utility.stringToBytes('Hey', TextEncoding.unicode),
        ...Utility.intToFourBytes(255),
        ...Utility.stringToBytes('Hey', TextEncoding.unicode),
        ...Utility.intToFourBytes(500),
        ...Utility.stringToBytes('Hey', TextEncoding.unicode),
        ...Utility.intToFourBytes(512),
      ];

      expect(bytes.sublist(0, 10), frameHeaderBytes);
      expect(bytes.sublist(10, 38), frameConfigBytes);
      expect(bytes.sublist(38), lyricsBytes);
    });

    test('toLrc() returns correct result 1', () {
      // Arrange

      // Act
      String srtLyrics = sampleSyltLyricsData.toLrc();

      // Assert
      String expectedResult = '[00:00,255]: Hey\n'
          '[00:00,500]: Hey\n'
          '[00:00,512]: Hey\n';
      expect(srtLyrics, expectedResult);
    });

    test('toSrt() returns correct result 1', () {
      // Arrange

      // Act
      String srtLyrics = sampleSyltLyricsData.toSrt();

      // Assert
      String expectedResult =
          '1\r\n00:00:00,255 --> 00:00:00,500\r\nHey\r\n\r\n'
          '2\r\n00:00:00,500 --> 00:00:00,512\r\nHey\r\n\r\n'
          '3\r\n00:00:00,512 --> 00:00:00,512\r\nHey\r\n\r\n';
      expect(srtLyrics, expectedResult);
    });
  });
}


lib/src/enums.dart
/// The [TextEncoding] determines how the text is encoded within the mp3 file.
///
/// The [TextEncoding] can be one byte per character which is indicated
/// by [TextEndcoding.iso8859_1] or two bytes per character when
/// [TextEncoding.unicode] is used.
enum TextEncoding {
  iso8859_1(rawByte: 0),
  unicode(rawByte: 1);

  final int rawByte;

  const TextEncoding({required this.rawByte});

  static fromRawByte(int rawByte) {
    switch (rawByte) {
      case 0:
        return TextEncoding.iso8859_1;
      case 1:
        return TextEncoding.unicode;
      default:
        throw ArgumentError();
    }
  }
}

/// The [TimestampFormat] determines how the timestap byte data is to be
/// interpreted.
///
/// The option [TimestampFormat.absTime32BitMpegFrames] indicates that the
/// timestamp is 4 bytes wide an stores the offset in time-units of mpeg frames.
/// The option [TimestampFormat.absTime32BitMilliseconds] also uses 4 bytes for
/// storing a timestamp but uses milliseconds as a unit instead.
enum TimestampFormat {
  absTime32BitMpegFrames(rawByte: 1),
  absTime32BitMilliseconds(rawByte: 2);

  final int rawByte;

  const TimestampFormat({required this.rawByte});

  static fromRawByte(int rawByte) {
    switch (rawByte) {
      case 1:
        return TimestampFormat.absTime32BitMpegFrames;
      case 2:
        return TimestampFormat.absTime32BitMilliseconds;
      default:
        throw ArgumentError();
    }
  }
}

/// The [ContentType] serves as a means to more clearly describe the data stored
/// within the SYLT-frame.
///
/// This enums defines all content types described within the mp3 standard.
enum ContentType {
  other(
    rawByte: 0,
    description: 'other',
  ),
  lyrics(
    rawByte: 1,
    description: 'lyrics',
  ),
  textTranscription(
    rawByte: 2,
    description: 'text transcription',
  ),
  movementOrPartName(
    rawByte: 3,
    description: 'movement/part name (e.g. "Adagio")',
  ),
  events(
    rawByte: 4,
    description: 'events (e.g. "Don Quijote enters the stage")',
  ),
  chord(
    rawByte: 5,
    description: 'chord (e.g. "Bb F Fsus")',
  ),
  triviaOrPopUpInformation(
    rawByte: 6,
    description: 'trivia/"pop up" information',
  );

  final int rawByte;
  final String description;

  const ContentType({required this.rawByte, required this.description});

  static fromRawByte(int rawByte) {
    switch (rawByte) {
      case 0:
        return ContentType.other;
      case 1:
        return ContentType.lyrics;
      case 2:
        return ContentType.textTranscription;
      case 3:
        return ContentType.movementOrPartName;
      case 4:
        return ContentType.events;
      case 5:
        return ContentType.chord;
      case 6:
        return ContentType.triviaOrPopUpInformation;
      default:
        throw ArgumentError();
    }
  }
}


lib/src/utility.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:sylt_parser/src/enums.dart';

class Utility {
  static List<int> languageStringToBytes(String languageString) {
    return languageString.codeUnits;
  }

  static String languageStringFromBytes(int byte1, int byte2, int byte3) {
    return String.fromCharCode(byte1) +
        String.fromCharCode(byte2) +
        String.fromCharCode(byte3);
  }

  static List<int> stringToBytes(
    String input,
    TextEncoding textEncoding, {
    bool nullTerminated = true,
  }) {
    List<int> bytes = [];

    List<int> codeUnits = input.codeUnits;

    // For single byte encoding just use the codepoints trimmed to 1 byte
    // if [TextEncoding.unicode] is used we transform the data to 2 bytes per
    // character instead
    List<int> textBytes = textEncoding == TextEncoding.iso8859_1
        ? codeUnits
        : Uint16List.fromList(codeUnits).buffer.asUint8List();

    bytes.addAll(textBytes);

    if (nullTerminated) {
      bytes.add(0);
      if (textEncoding == TextEncoding.unicode) {
        bytes.add(0);
      }
    }

    return bytes;
  }

  static String stringFromBytes(List<int> bytes, TextEncoding textEncoding,
      {bool nullTerminated = true}) {
    if (nullTerminated) {
      int nullTerminationBytes = textEncoding == TextEncoding.iso8859_1 ? 1 : 2;
      bytes = bytes.sublist(0, bytes.length - nullTerminationBytes);
    }

    if (textEncoding == TextEncoding.iso8859_1) {
      return utf8.decode(bytes);
    } else {
      return String.fromCharCodes(
          Uint8List.fromList(bytes).buffer.asUint16List());
    }
  }

  static int fromSyncEncoding(int i1, int i2, int i3, int i4) {
    return i1 << 21 | i2 << 14 | i3 << 7 | i4;
  }

  static int fourBytesToInt(int b1, int b2, int b3, int b4) {
    return b1 << 24 | b2 << 16 | b3 << 8 | b4;
  }

  static List<int> intToFourBytes(int number) {
    int byte1 = number >> 24;
    int byte2 = (number >> 16) & 255;
    int byte3 = (number >> 8) & 255;
    int byte4 = number & 255;

    return [byte1, byte2, byte3, byte4];
  }

  static String timestampInMsToString(int timeStampInMilliSeconds) {
    int timeStampInSeconds = timeStampInMilliSeconds ~/ 1000;
    int timeStampInMinutes = timeStampInSeconds ~/ 60;

    String milliSeconds =
        (timeStampInMilliSeconds % 1000).toString().padLeft(3, '0');
    String seconds = (timeStampInSeconds % 60).toString().padLeft(2, '0');
    String minutes = (timeStampInMinutes % 60).toString().padLeft(2, '0');
    String formattedTimeStamp = '$minutes:$seconds,$milliSeconds';

    return formattedTimeStamp;
  }
}


lib/src/sylt_lyrics_data.dart
import 'enums.dart';
import 'utility.dart';

class SyltLyricsData {
  final TextEncoding textEncoding;
  final String language;
  final TimestampFormat timestampFormat;
  final ContentType contentType;
  final String contentDescriptor;
  final List<(String, int)> lyricsLines;

  SyltLyricsData({
    required this.textEncoding,
    required this.language,
    required this.timestampFormat,
    required this.contentType,
    required this.contentDescriptor,
    required this.lyricsLines,
  });

  SyltLyricsData copyWith({
    TextEncoding? textEncoding,
    String? language,
    TimestampFormat? timestampFormat,
    ContentType? contentType,
    String? contentDescriptor,
    List<(String, int)>? lyricsLines,
  }) {
    return SyltLyricsData(
        textEncoding: textEncoding ?? this.textEncoding,
        language: language ?? this.language,
        timestampFormat: timestampFormat ?? this.timestampFormat,
        contentType: contentType ?? this.contentType,
        contentDescriptor: contentDescriptor ?? this.contentDescriptor,
        lyricsLines: lyricsLines ?? this.lyricsLines);
  }

  List<int> toBytes() {
    int s = 83; // S
    int y = 89; // Y
    int l = 76; // L
    int t = 84; // T

    List<int> frameHeaderIdBytes = [s, y, l, t];
    List<int> frameHeaderSizeBytes = [0, 0, 0, 0];
    List<int> frameHeaderFlagsBytes = [0, 0];

    List<int> contentDescriptorBytes =
        Utility.stringToBytes(contentDescriptor, textEncoding);

    List<int> lyricsBytes = [];

    for ((String, int) lineData in lyricsLines) {
      lyricsBytes.addAll(Utility.stringToBytes(lineData.$1, textEncoding));
      lyricsBytes.addAll(Utility.intToFourBytes(lineData.$2));
    }

    int frameSize = 6 + contentDescriptorBytes.length + lyricsBytes.length;
    frameHeaderSizeBytes = Utility.intToFourBytes(frameSize);

    return [
      ...frameHeaderIdBytes,
      ...frameHeaderSizeBytes,
      ...frameHeaderFlagsBytes,
      textEncoding.rawByte,
      ...Utility.languageStringToBytes(language),
      timestampFormat.rawByte,
      contentType.rawByte,
      ...contentDescriptorBytes,
      ...lyricsBytes,
    ];
  }

  /// Convert to format
  /// 1\r\n00:00:01,000 --> 00:00:29,950\r\nHere goes the lyrics\r\n\r\n...
  static String syltLyricsToSrtLyrics(List<String> syltLyrics) {
    int counter = 1;
    String lastTimestamp = '00:00:00,000';
    String lyrics = '';
    for (String line in syltLyrics) {
      String timeStamp = '00:${line.substring(1, 10)}';
      String text = line.substring(12, line.length);
      String newLine =
          '$counter\r\n$lastTimestamp --> $timeStamp\r\n$text\r\n\r\n';
      lyrics += newLine;

      lastTimestamp = timeStamp;
      counter++;
    }

    return lyrics;
  }

  String toLrc() {
    String lyrics = '';

    for ((String, int) lineData in lyricsLines) {
      String timestamp = Utility.timestampInMsToString(lineData.$2);
      lyrics += '[$timestamp]: ${lineData.$1}\n';
    }

    return lyrics;
  }

  String toSrt() {
    String lyrics = '';

    for (int i = 0; i < lyricsLines.length; i++) {
      String timestamp = Utility.timestampInMsToString(lyricsLines[i].$2);

      String nextTimestamp = '';
      if (i + 1 == lyricsLines.length) {
        nextTimestamp = Utility.timestampInMsToString(
            lyricsLines[lyricsLines.length - 1].$2);
      } else {
        nextTimestamp = Utility.timestampInMsToString(lyricsLines[i + 1].$2);
      }

      String lyricsLine = lyricsLines[i].$1;

      // If the line is empty than fill it with a blank space, in order to not
      // have any blank lines apart from those at the end of each SRT lyrics
      // line. This helps to avoid errors made by parsers of this format.
      if (lyricsLine.isEmpty) {
        lyricsLine = ' ';
      }

      String newLine =
          '${i + 1}\r\n00:$timestamp --> 00:$nextTimestamp\r\n$lyricsLine\r\n\r\n';
      lyrics += newLine;
    }

    return lyrics;
  }
}


example/sylt_parser_example.dart
import 'package:sylt_parser/sylt_parser.dart';

void main() {
  // Create a sample SyltLyricsData object and convert it to bytes. This list of
  // bytes would normally be loaded from an mp3 file.
  List<int> mp3Bytes = SyltLyricsData(
    textEncoding: TextEncoding.iso8859_1,
    language: 'eng',
    timestampFormat: TimestampFormat.absTime32BitMilliseconds,
    contentType: ContentType.lyrics,
    contentDescriptor: 'MiniLyrics',
    lyricsLines: [
      ('Hey', 255),
      ('Hey', 500),
      ('Hey', 512),
    ],
  ).toBytes();

  // Parse the list of bytes coming from the mp3 file.
  // The library searches for the SYLT-frame within the ID3 tag and parses it,
  // if one such frame is found.
  SyltLyricsData lyricsData =
      SyltLyricsFromMp3Parser.parseMp3BytesToSyltLyricsData(mp3Bytes);

  // The intermediate data can be parsed to either LRC or SRT lyrics formats:
  String lrcLyrics = lyricsData.toLrc();
  print('LRC lyrics:\n$lrcLyrics');

  String srtLyrics = lyricsData.toSrt();
  print('SRT lyrics:\n$srtLyrics');
}


lib/src/sylt_lyrics_from_mp3_parser.dart
import 'package:sylt_parser/src/enums.dart';

import 'exceptions.dart';
import 'sylt_lyrics_data.dart';
import 'utility.dart';

/// This class provides static functionality for parsing the SYLT data within
/// mp3 files into an object containing the time-coded lyrcis data.
class SyltLyricsFromMp3Parser {
  /// Parses the time-coded lyrics data within an mp3 file represented as a byte
  /// list.
  ///
  /// The whole file is search for the startdata that indicates the beginning
  /// of a SYLT frame. When the starting offset is found the data in the frame
  /// is parsed into a [SyltLyricsData] object.
  static SyltLyricsData parseMp3BytesToSyltLyricsData(List<int> mp3Bytes) {
    int syltFrameStartingOffset = _findSyltFrameStartingOffset(mp3Bytes);

    int cursor = syltFrameStartingOffset;
    cursor += 4; // Move to frame size offset
    int frameSize = Utility.fourBytesToInt(
      mp3Bytes[cursor],
      mp3Bytes[cursor + 1],
      mp3Bytes[cursor + 2],
      mp3Bytes[cursor + 3],
    );

    // Add 10 for the 10 bytes of frame header not accounted in the frameSize
    int frameEnd = syltFrameStartingOffset + 10 + frameSize;

    // for (int i = syltFrameStartingOffset; i < frameEnd; i++) {
    //   debugPrint(mp3Bytes[i].toString());
    //   debugPrint(
    //       'C${i.toString().padLeft(3, ' ')}: (${String.fromCharCode(mp3Bytes[i])}) or ${mp3Bytes[i]}');
    // }

    cursor += 4; // Skip frame size
    cursor += 2; // Skip frame header flags

    TextEncoding textEncoding = TextEncoding.fromRawByte(mp3Bytes[cursor]);
    cursor += 1; // Skip text encoding

    String languageString = Utility.languageStringFromBytes(
      mp3Bytes[cursor],
      mp3Bytes[cursor + 1],
      mp3Bytes[cursor + 2],
    );
    cursor += 3; // Skip language code

    TimestampFormat timestampFormat =
        TimestampFormat.fromRawByte(mp3Bytes[cursor]);
    cursor += 1; // Skip time stamp format

    ContentType contentType = ContentType.fromRawByte(mp3Bytes[cursor]);
    cursor += 1; // Skip content type

    // Skip all chars of the content descriptor string until we find the null
    // termination character
    String contentDescriptor = '';
    while (mp3Bytes[cursor] != 0) {
      contentDescriptor += String.fromCharCode(mp3Bytes[cursor]);
      cursor++;
      if (textEncoding == TextEncoding.unicode) {
        cursor++;
      }
    }
    cursor++; // Skip content descriptor string null termination character
    if (textEncoding == TextEncoding.unicode) {
      cursor++;
    }

    List<(String, int)> lyricsLines = [];
    List<int> currentLineBytes = [];
    while (cursor < frameEnd) {
      if (mp3Bytes[cursor] == 0 &&
          (textEncoding == TextEncoding.iso8859_1 ||
              mp3Bytes[cursor + 1] == 0)) {
        cursor++; // Skip lyrics line null termination character
        if (textEncoding == TextEncoding.unicode) {
          cursor++;
        }
        for (int i = 0; i < 4; i++) {
          currentLineBytes.add(mp3Bytes[cursor]);
          cursor++;
        }
        lyricsLines.add(_lyricsLineFromBytes(
          currentLineBytes,
          textEncoding,
          nullTerminated: false,
        ));
        currentLineBytes = [];
      } else {
        currentLineBytes.add(mp3Bytes[cursor]);
        cursor++;
        if (textEncoding == TextEncoding.unicode) {
          currentLineBytes.add(mp3Bytes[cursor]);
          cursor++;
        }
      }
    }

    return SyltLyricsData(
        textEncoding: textEncoding,
        language: languageString,
        timestampFormat: timestampFormat,
        contentType: contentType,
        contentDescriptor: contentDescriptor,
        lyricsLines: lyricsLines);
  }

  static (String, int) _lyricsLineFromBytes(
    List<int> bytes,
    TextEncoding textEncoding, {
    bool nullTerminated = true,
  }) {
    List<int> textBytes = bytes.sublist(0, bytes.length - 4);

    String lyricsLine = Utility.stringFromBytes(
      textBytes,
      textEncoding,
      nullTerminated: nullTerminated,
    );

    List<int> timestampBytes = bytes.sublist(bytes.length - 4);
    int timestamp = Utility.fourBytesToInt(
      timestampBytes[0],
      timestampBytes[1],
      timestampBytes[2],
      timestampBytes[3],
    );

    return (lyricsLine, timestamp);
  }

  /// Determine the byte offset where the sylt header starts
  static int _findSyltFrameStartingOffset(List<int> mp3Bytes) {
    int s = 83; // S
    int y = 89; // Y
    int l = 76; // L
    int t = 84; // T

    int? syltHeaderStartingOffset;

    for (int i = 0; i < mp3Bytes.length - 3; i++) {
      if (mp3Bytes[i] == s &&
          mp3Bytes[i + 1] == y &&
          mp3Bytes[i + 2] == l &&
          mp3Bytes[i + 3] == t) {
        syltHeaderStartingOffset = i;
        break;
      }
    }

    if (syltHeaderStartingOffset == null) {
      throw NoSyltFrameFoundException();
    }

    return syltHeaderStartingOffset;
  }
}
